src/main/java/minegame159/meteorclient/systems/modules/render/WaypointsModule.java

```diff
@@ -33,16 +35,104 @@
+import net.minecraft.text.LiteralText;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.Vec3d;
 
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.ListIterator;
+
+import static minegame159.meteorclient.utils.player.ChatUtils.formatCoords;
+
 public class WaypointsModule extends Module {
     private static final Color GRAY = new Color(200, 200, 200);
+    
+    private final SettingGroup sgDeathPosition = settings.createGroup("death-position");
+
+    private final Setting<Integer> dpWpAmouunt = sgDeathPosition.add(new IntSetting.Builder()
+            .name("wp-amount")
+            .description("The amount of death positions to save, 0 to disable")
+            .min(0)
+            .sliderMin(0)
+            .sliderMax(20)
+            .defaultValue(0)
+            .onChanged(this::cleanDeathWPs)
+            .build()
+    );
+
+    private final Setting<Boolean> dpChat = sgDeathPosition.add(new BoolSetting.Builder()
+            .name("chat")
+            .description("Send a chat message with your position once you die")
+            .defaultValue(false)
+            .build()
+    );
 
     public WaypointsModule() {
         super(Categories.Render, "waypoints", "Allows you to create waypoints.");
     }
 
+    private final SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm:ss");
+
+    @EventHandler
+    private void onOpenScreen(OpenScreenEvent event) {
+        if (!(event.screen instanceof DeathScreen)) return;
+        if (mc.player == null) return;
+        Vec3d dmgPos = mc.player.getPos();
+
+        String time = dateFormat.format(new Date());
+        if (dpChat.get()) {
+            BaseText text = new LiteralText("Died at ");
+            text.append(formatCoords(dmgPos));
+            text.append(String.format(" on %s.", time));
+            info(text);
+        }
+
+        // Create waypoint
+        if (dpWpAmouunt.get() > 0) {
+            Waypoint waypoint = new Waypoint();
+            waypoint.name = "Death " + time;
+            waypoint.icon = "skull";
+            waypoint.x = (int) dmgPos.x;
+            waypoint.y = (int) dmgPos.y + 2;
+            waypoint.z = (int) dmgPos.z;
+            waypoint.maxVisibleDistance = Integer.MAX_VALUE;
+            waypoint.actualDimension = PlayerUtils.getDimension();
+
+            switch (waypoint.actualDimension) {
+                case Overworld:
+                    waypoint.overworld = true;
+                    break;
+                case Nether:
+                    waypoint.nether = true;
+                    break;
+                case End:
+                    waypoint.end = true;
+                    break;
+            }
+
+            Waypoints.get().add(waypoint);
+        }
+        cleanDeathWPs(dpWpAmouunt.get());
+    }
+
+    private void cleanDeathWPs(int max) {
+        int oldWpC = 0;
+
+        ListIterator<Waypoint> wps = Waypoints.get().iteratorReverse();
```

Comment:
Can't it just use normal iterator that goes from the start?

Reply:
The point is to keep the most recent waypoints and remove earlier ones. The normal iterator starts with older ones and new ones would be removed. In order to use the normal iterator I would need to do 2 iterations, one to find all death waypoints, then get the older ones from a list generated from that and finally iterate through that list to delete the older waypoints. This is simpler

Reply:
Correct me if I am wrong but the oldest waypoints are at the start of the list. So wouldn't it make sense to loop from the start to remove the oldest first?

Reply:
I want to keep a number of death waypoints so I need to know how many death waypoints will follow the current waypoint if it is a death waypoint.

Reply:
Oh right, sorry.